src/types/token_amount.ts

```diff
@@ -0,0 +1,58 @@
+// Utils
+import { BigNumber } from "../../utils/bignumber";
+
+// Types
+import { Token } from "../types";
+
+export class TokenAmount {
+    public static fromRawAmount(rawAmount: BigNumber, symbol: string): TokenAmount {
+        return new TokenAmount(rawAmount, new Token(symbol));
+    }
+
+    public static fromDecimalAmount(decimalAmount: BigNumber, symbol: string) {
+        const token = new Token(symbol);
+        const rawAmount = TokenAmount.convertToRaw(decimalAmount, token.decimals);
+
+        return new TokenAmount(rawAmount, token);
+    }
+
+    private static convertToRaw(decimalAmount: BigNumber, numDecimals: BigNumber): BigNumber {
+        return decimalAmount.mul(new BigNumber(10).pow(numDecimals.toNumber()));
+    }
+
+    private static convertToDecimal(rawAmount: BigNumber, numDecimals: BigNumber): BigNumber {
+        return rawAmount.div(new BigNumber(10).pow(numDecimals.toNumber()));
+    }
+
+    public readonly rawAmount: BigNumber;
+    private token: Token;
+
+    private constructor(rawAmount: BigNumber, token: Token) {
+        this.rawAmount = rawAmount;
+        this.token = token;
+    }
+
+    get address(): string {
```

Comment:
For these methods that are delegating to the `token` instance, I think it would be clearer if they were prefixed, I.E. `tokenAmount.tokenAddress()` => // the address of the token, rather than `tokenAmount.address()` which is less clear.

Reply:
Especially given `tokenAmount.decimals()` which is definitely confusing.